### PR TITLE
fix(devcontainer): refuse to open from a git worktree

### LINKED ISSUE #593: fix(devcontainer): refuse to open from a git worktree (fail fast instead of mid-postCreate)
## Problem

`devcontainer up` fails partway through provisioning when the workspace folder is a linked git worktree.

Reproduction (observed during Level 1 verification of #592):

```bash
devcontainer up \
  --workspace-folder /home/<user>/<repo>/.worktrees/<name> \
  --config .devcontainer/cpu/devcontainer.json
```

The container is created successfully, the image layer and mounts are correct, but `postCreateCommand` (`.devcontainer/post-create.sh`) exits 128 with:

```
fatal: not a git repository: /home/<user>/<repo>/.git/worktrees/<name>
Error: Command failed: /bin/sh -c .devcontainer/post-create.sh
```

The container exists but is half-configured: `pre-commit install` didn't run, `git config --global --add safe.directory` didn't run, and any in-container `git` command will fail. Not a caller error — every worktree user hits this.

## Root cause

A linked worktree's `.git` is a **pointer file**, not a directory:

```
$ cat .worktrees/<name>/.git
gitdir: /home/<user>/<repo>/.git/worktrees/<name>
```

That's an **absolute host path** into the parent repo's admin directory. The devcontainer bind-mounts only the worktree itself (via `workspaceMount`), so the path the pointer references doesn't exist inside the container. Git opens `.git`, reads the pointer, tries to stat the target, and fails.

This is structural. No amount of tweaking inside the current container produces a working git, because the data git needs is on the host outside the mounted area.

### Option A — Silent degrade in `post-create.sh`

Detect the unusable git early and `exit 0` cleanly instead of erroring out.

**Pros**
- Smallest change (3-line guard).
- Container boots, is usable for build/test work.
- Preserves worktree-in-devcontainer as a workflow (with caveats).

**Cons**
- Failure is invisible post-provisioning. `git status` inside the container will fail much later with a confusing message.
- `pre-commit install` silently didn't run; commits from inside the container skip the project's hooks.
- Creates a "half-configured container" state not reflected anywhere observable (no `docker inspect` field, no banner).
- Docs become load-bearing; easy to forget this limitation exists.

### Option A+tripwire — Silent skip plus interactive warning

Same as A, but leave a `$HOME/.devcontainer-worktree-warning` file and source it from `~/.bashrc` so every interactive shell prints the limitation.

**Pros**
- Visible-on-every-shell degrade, not just at provision time.
- Preserves worktree-in-devcontainer.

**Cons**
- Still half-configured; now just with a warning banner.
- Extra state to maintain (warning file, bashrc edit).

### Option B — Hard fail in `initializeCommand`

Detect the worktree pointer **before** the container is built; exit with a clear message naming the two supported workflows.

**Pros**
- Honest contract: the devcontainer assumes git works. If it can't, refuse.
- Fast feedback (seconds, not after a minutes-long container build).
- No half-configured container states.
- Error message is the docs.

**Cons**
- Blocks one specific workflow: running **parallel devcontainers** for parallel branches by opening each worktree in its own devcontainer. Narrow but legitimate (A/B comparisons, keeping a repro alongside a feature branch).
- Dev who wants that workflow has to know it needs local edits (no built-in opt-in).

### Option C — Conditional parent-`.git` bind-mount

Detect worktree in `initializeCommand`, conditionally bind-mount the parent repo's `.git` at its exact absolute host path inside the container (so the pointer resolves).

**Pros**
- Actually fixes git inside the container. Worktree-in-devcontainer becomes a full workflow.

**Cons**
- `devcontainer.json` doesn't let `initializeCommand` inject conditional mounts cleanly. Requires either a second opt-in config or migrating to docker-compose with an override file.
- Mount target must be an absolute host path (per-user, per-machine), so config relies on layout conventions or runtime computation.
- Parking this in a follow-up is fine; don't block the current bug on it.

## Decision

**Option B — hard fail in `initializeCommand`, escape hatch documented in code comments only.**

Rationale:
- `CLAUDE.md` already supports two valid branch-isolation workflows that don't require the worktree+devcontainer combo: (1) host-only with worktree, (2) devcontainer-from-root with `git checkout -B`. We're not closing off branch isolation — just closing off one specific combination.
- The half-state from Option A is worse than a clean refusal.
- Building opt-in machinery (env var, separate config) for a narrow edge case violates YAGNI. Someone who genuinely needs parallel-branch devcontainers can patch locally; the escape hatch lives in code comments next to the check, not in committed support code.

Option C remains as a proper future fix; file as follow-up.

## Minimal spec

New file `.devcontainer/initialize.sh`:

- Keeps existing behavior (`[ -f .env ] || touch .env`).
- Detects `.git` as a pointer file (starts with `gitdir:`); on match, prints a multi-line error to stderr naming the two supported workflows and exits 1.
- Inline comment above the check explains the parallel-devcontainer edge case: to use it, comment out this block *and* the `git config` / `pre-commit install` lines in `post-create.sh`; do not commit the edits.

Updates to both `.devcontainer/cpu/devcontainer.json` and `.devcontainer/gpu/devcontainer.json`:

- Replace `"initializeCommand": "[ -f .env ] || touch .env"` with `"initializeCommand": "bash .devcontainer/initialize.sh"`.

No changes to `post-create.sh`, no changes to the Dockerfile, no new env vars, no new config files.

### Verification

- Run `devcontainer up --workspace-folder <worktree>` and confirm: exit code non-zero, clear error to stderr, no container created.
- Run `devcontainer up --workspace-folder <main-repo-root>` and confirm: container builds, `postCreateCommand` runs to completion as before.
- Open a shell in the main-repo devcontainer and confirm `pre-commit install` ran (`.git/hooks/pre-commit` exists) and `git status` works.

### Follow-up

File a Task under #153 for Option C if worktree-in-devcontainer becomes a priority workflow later.

### PR DESCRIPTION
**Pull request by ktinubu**
## Summary

- Moves the devcontainer's host-side `initializeCommand` from an inline shell string into a new `.devcontainer/initialize.sh`.
- The script detects a git worktree pointer file and hard-fails before any image build, with a clear error naming the two supported branch-isolation workflows.
- The parallel-devcontainer edge case is documented inline as a local-only patch — no committed machinery.

Fixes #593

## Test plan

- [ ] From a git worktree, run `devcontainer up --workspace-folder . --config .devcontainer/cpu/devcontainer.json`. Expect non-zero exit, multi-line error to stderr, and **no container or image built**.
- [ ] From the main repo root, run the same command. Expect it to proceed past `initializeCommand` and build/provision normally (including `pre-commit install` succeeding).
- [ ] From the main repo, confirm `.devcontainer/initialize.sh` still creates a `.env` file if absent (preserves the original responsibility).
- [ ] Confirm the GPU config (`--config .devcontainer/gpu/devcontainer.json`) inherits the same hardstop.

### CODE REVIEW

**Comment by ktinubu** [comment]
## Verification Results

Branch: `fix/devcontainer-worktree-hardstop` at `e35557305479ea527d3adea4f6560c4b08853fa9`.

- [x] **`devcontainer up` from a git worktree hard-fails in `initializeCommand` before any image build, with a readable error** *(Level 1 — real devcontainer CLI, real initializeCommand, real exit semantics.)*

  ```bash
  $ devcontainer up --workspace-folder . --config .devcontainer/cpu/devcontainer.json 2>&1 | grep -E '^(ERROR|The workspace|  1\.|  2\.|See \.|Error: Command failed)'
  The workspace's .git is a pointer file referencing a host path that is not
    1. Host-only:    open <worktree>/ in a host editor, skip the devcontainer.
    2. Devcontainer: open the main repo in the devcontainer, then
  See .devcontainer/initialize.sh for the parallel-devcontainer escape hatch.
  Error: Command failed: /bin/sh -c bash .devcontainer/initialize.sh

  $ devcontainer up ... > /dev/null 2>&1; echo "exit=$?"
  exit=1

  $ docker image ls --format '{{.Repository}}' | grep -c fix-devcontainer-worktree-hardstop
  0
  $ docker ps -a --filter label=devcontainer.local_folder=<worktree> --format '{{.ID}}' | wc -l
  0
  ```
  Zero images and zero containers — hardstop aborts the CLI before Docker does any work. (The grep snips the multi-line "ERROR:" header and the "`git checkout -B <branch>`" line to keep this block tight; the full error was captured verbatim in the implementation test output.)

- [x] **From a non-worktree workspace (fresh clone, `.git` is a real directory), `initializeCommand` exits 0 and still touches `.env`** *(Level 2 — ran the actual script the CLI invokes, with the realistic cwd and filesystem shape. Level 1 would add a multi-minute `devcontainer up` build for no new signal: the CLI's only contribution on this path is to invoke the script, and that wrapper's correctness isn't this PR's promise.)*

  ```bash
  $ TMP=$(mktemp -d) && cd "$TMP" && git clone --depth 1 --branch fix/devcontainer-worktree-hardstop https://github.com/tinaudio/synth-setter.git clone && cd clone
  $ stat -c %F .git
  directory
  $ bash .devcontainer/initialize.sh; echo "exit=$?"
  exit=0
  $ ls -l .env
  -rw-rw-r-- 1 k-linux k-linux 0 ... .env
  ```
  Confirms: no false positive on main-line clones, and the pre-existing `.env` touch responsibility is preserved.

- [x] **GPU config (`.devcontainer/gpu/devcontainer.json`) inherits the same hardstop** *(Level 1 — ran the CLI against the GPU config; the hardstop fires in `initializeCommand` before any `--gpus all` machinery is needed, so no NVIDIA runtime is required to test it.)*

  ```bash
  $ devcontainer up --workspace-folder . --config .devcontainer/gpu/devcontainer.json 2>&1 | grep -E '^(ERROR|The workspace|  1\.|  2\.|See \.|Error: Command failed)'
  The workspace's .git is a pointer file referencing a host path that is not
    1. Host-only:    open <worktree>/ in a host editor, skip the devcontainer.
    2. Devcontainer: open the main repo in the devcontainer, then
  See .devcontainer/initialize.sh for the parallel-devcontainer escape hatch.
  Error: Command failed: /bin/sh -c bash .devcontainer/initialize.sh
  $ devcontainer up ... > /dev/null 2>&1; echo "exit=$?"
  exit=1
  ```

- [x] **Both configs reference the same `initializeCommand` script** *(Level 5 — static symmetry check, justified: the runtime behavior is identical because they run the same script. A Level 1 replay on each would re-prove what Check 1 and Check 3 already proved.)*

  ```bash
  $ python3 -c "
  import json
  for p in ['.devcontainer/cpu/devcontainer.json', '.devcontainer/gpu/devcontainer.json']:
      cmd = json.load(open(p))['initializeCommand']
      print(f'{p}: {cmd}')
  "
  .devcontainer/cpu/devcontainer.json: bash .devcontainer/initialize.sh
  .devcontainer/gpu/devcontainer.json: bash .devcontainer/initialize.sh
  ```

### Summary

All four checks pass. The hardstop fires at the correct gate (pre-build), with a clear error, for both configs; the non-worktree happy path is unaffected; pre-existing `.env` behavior is preserved. Safe to merge.

File: .devcontainer/initialize.sh [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+# EDGE CASE — parallel devcontainers for parallel branches:
+# If you genuinely need to run two devcontainers on two worktrees simultaneously
+# (e.g. to A/B-test behavior between branches without rebuilding), comment out
+# the worktree check below AND the `git config` / `pre-commit install` lines in
+# .devcontainer/post-create.sh. DO NOT commit those edits.
+
+set -euo pipefail
+
+[ -f .env ] || touch .env
+
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply .devcontainer/initialize.sh]: Fixed in `0ee4881`. Moved `[ -f .env ] || touch .env` below the pointer-file guard so an aborted run no longer leaves a stray empty `.env` in the workspace. Updated the top-of-file comment to reflect the new order.

File: .devcontainer/initialize.sh [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+if [ -f .git ]; then
+  first_line=""
+  read -r first_line < .git 2>/dev/null || true
+  case "$first_line" in
+    gitdir:*)
+      cat >&2 <<'ERR'
+ERROR: This devcontainer cannot be opened from a git worktree.
+
+The workspace's .git is a pointer file referencing a host path that is not
+mounted into the container. Supported workflows:
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply .devcontainer/initialize.sh]: Fixed in `0ee4881`. Kept the guard as-is (the `gitdir:*` match is actually correct — worktrees, submodules, and `git clone --separate-git-dir` repos all have the same failure mode: pointer target outside the workspaceMount). Broadened the error wording from "worktree" to "workspace whose `.git` is a pointer file" and named worktree as the common case. The rare non-worktree hits will now read correctly.
